inc/templates.php

```diff
@@ -274,3 +279,212 @@ function convert_blocks_to_components( array $blocks ): array {
+	}
+
+	return [
+		'name'          => (string) ( $component['name'] ?? '' ),
+		'config'        => (array) ( $component['config'] ?? [] ),
+		'data_provider' => (array) ( $component['data_provider'] ?? [] ),
+		'children'      => (array) ( $component['children'] ?? [] ),
+	];
+}
+
+/**
+ * Modify the final output of a component.
+ *
+ * @param array $component Component.
+ * @return array
+ */
+function validate_final_component( $component ) {
+
+	if ( is_string( $component ) ) {
+		return $component;
+	}
+
+	$component                  = ensure_component_fields_exist( $component );
+	$component['config']        = (object) $component['config'];
+	$component['data_provider'] = (object) $component['data_provider'];
+	$component['children']      = array_values( array_filter( (array) ( $component['children'] ?? [] ) ) );
+
+	unset( $component['data_provider'] );
+
+	return $component;
+}
+
+
+/**
+ * Pull in template parts.
+ *
+ * @param array $component Component.
+ * @return array
+ */
+function handle_template_parts( $component ) {
+
+	// Check if this is a template part.
+	$name = $component['name'] ?? '';
+	if ( 0 !== strpos( $name, 'template-parts/' ) ) {
+		return $component;
+	}
+
+	$template_part_name = str_replace( 'template-parts/', '', $component['name'] );
+
+	$template = \WP_Irving\Templates\locate_template( [ $template_part_name ], '/template-parts/' );
+
+	$template_data = \WP_Irving\Templates\prepare_data_from_template( $template );
+
+	if ( isset( $template_data['name'] ) ) {
+		$template_data = [ $template_data ];
+	}
+
+	$component['name'] = 'irving/passthrough';
+	$component['children'] = $template_data;
+
+	return $component;
+}
+
+/**
+ * Pass data provider values down to children components.
+ *
+ * @param array $component Component.
+ * @return array
+ */
+function handle_data_provider( $component ) {
+
+	// If there's no data provider, or children, don't do anything.
+	if ( empty( $component['data_provider'] ) || empty( $component['children'] ) ) {
+		return $component;
+	}
+
+	foreach ( $component['children'] as &$child_component ) {
+
+		if ( is_string( $child_component ) ) {
+			continue;
+		}
+
+		$child_component = ensure_component_fields_exist( $child_component );
+
+		$child_component['data_provider'] = array_merge_recursive(
+			$child_component['data_provider'],
+			$component['data_provider']
+		);
+	}
+
+	return $component;
+}
+
+/**
+ * Loop through component config values. If we match a {{component/name}} then
+ * we execute the callback, and return the value to the config.
+ *
+ * @todo Determine if this functionality can be scrapped in favor of a data
+ *       provider. This is proof of concept purely for discussion and
+ *       consideration.
+ *
+ * @example
+ * {
+ *   "name": "example",
+ *   "config": {
+ *     "href": "{{post/permalink}}"
+ *   }
+ * }
+ *
+ * @param array $component Component.
+ * @return array
+ */
+function handle_component_config_callbacks( $component ): array {
```

Comment:
This is super helpful to see how you were thinking about this. I'd still prefer to find a way of handling this without introducing non-JSON-y concepts if possible but am ok keeping this as experimental for now while we're playing around unless you're ok with pulling this out and making this a follow-up PR as we're messing with a context API. Your call.

Reply:
I'd like to include it for now and trash it after the context stuff is setup, it's useful as a temporary solution so we're not blocked on experiments.